File: sdk/php/src/Internal/Infra/CallbackManager.php
<?php

namespace KuCoin\UniversalSDK\Internal\Infra;

use KuCoin\UniversalSDK\Internal\Interfaces\WebSocketMessageCallback;
use KuCoin\UniversalSDK\Internal\Utils\SubInfo;

class CallbackManager
{
    /** @var array<string, array<string, bool>> */
    private $idTopicMapping = [];

    /** @var array<string, Callback> */
    private $topicCallbackMapping = [];

    /** @var string */
    private $topicPrefix;

    public function __construct($topicPrefix)
    {
        $this->topicPrefix = $topicPrefix;
    }

    public function isEmpty()
    {
        return empty($this->idTopicMapping) && empty($this->topicCallbackMapping);
    }

    /**
     * @return SubInfo[]
     */
    public function getSubInfo()
    {
        $subInfoList = [];

        foreach ($this->idTopicMapping as $topics) {
            $info = new SubInfo($this->topicPrefix, [], null);

            foreach (array_keys($topics) as $topic) {
                $parts = explode(':', $topic);
                if (count($parts) === 2 && $parts[1] !== 'all') {
                    $info->args[] = $parts[1];
                }

                if (isset($this->topicCallbackMapping[$topic])) {
                    $info->callback = $this->topicCallbackMapping[$topic]->callback;
                }
            }

            $subInfoList[] = $info;
        }

        return $subInfoList;
    }

    public function add(SubInfo $subInfo): bool
    {
        $id = $subInfo->toId();
        if (isset($this->idTopicMapping[$id])) {
            return false;
        }

        $topicMap = [];

        foreach ($subInfo->topics() as $topic) {
            if (isset($this->topicCallbackMapping[$topic])) {
                continue;
            }

            $topicMap[$topic] = true;
            $this->topicCallbackMapping[$topic] = new Callback($subInfo->callback, $id, $topic);
        }

        $this->idTopicMapping[$id] = $topicMap;
        return true;
    }

    public function remove($id)
    {
        if (!isset($this->idTopicMapping[$id])) {
            return;
        }

        foreach ($this->idTopicMapping[$id] as $topic => $_) {
            unset($this->topicCallbackMapping[$topic]);
        }

        unset($this->idTopicMapping[$id]);
    }

    /**
     * @param string $topic
     * @return WebSocketMessageCallback|null
     */
    public function get($topic)
    {
        return isset($this->topicCallbackMapping[$topic])
            ? $this->topicCallbackMapping[$topic]->callback
            : null;
    }
}

class Callback
{
    /**
     * @var WebSocketMessageCallback $callback
     */
    public $callback;
    /**
     * @var string $id
     */
    public $id;
    /**
     * var string $topic
     */
    public $topic;

    /**
     * @param WebSocketMessageCallback $callback
     * @param string $id
     * @param $topic
     */
    public function __construct(WebSocketMessageCallback $callback, string $id, $topic)
    {
        $this->callback = $callback;
        $this->id = $id;
        $this->topic = $topic;
    }

}

File: sdk/java/src/main/java/com/kucoin/universal/sdk/internal/infra/CallbackManager.java
package com.kucoin.universal.sdk.internal.infra;

import com.kucoin.universal.sdk.internal.interfaces.WebSocketMessageCallback;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import lombok.Value;
import lombok.extern.slf4j.Slf4j;

@Slf4j
final class CallbackManager {

  /** id → topics set */
  private final ConcurrentMap<String, Set<String>> idTopics = new ConcurrentHashMap<>();

  /** topic → callback */
  private final ConcurrentMap<String, Callback> topicCbs = new ConcurrentHashMap<>();

  private final String topicPrefix;

  CallbackManager(String prefix) {
    this.topicPrefix = prefix;
  }

  boolean empty() {
    return idTopics.isEmpty() && topicCbs.isEmpty();
  }

  boolean add(SubInfo s) {
    // if id already exists -> duplicated subscribe
    Set<String> topicSet = s.topics();
    Set<String> existed = idTopics.putIfAbsent(s.toId(), topicSet);
    if (existed != null) {
      log.trace("callback exists id=[{}]", s.toId());
      return false;
    }
    // register every topic
    for (String t : topicSet) {
      topicCbs.putIfAbsent(t, new Callback(s.getCallback(), s.toId(), t));
    }
    return true;
  }

  void remove(String id) {
    Set<String> topics = idTopics.remove(id);
    if (topics == null) return;
    topics.forEach(topicCbs::remove);
  }

  WebSocketMessageCallback<?> get(String topic) {
    Callback cb = topicCbs.get(topic);
    return cb == null ? null : cb.callback;
  }

  List<SubInfo> getSubInfo() {
    List<SubInfo> list = new ArrayList<>();

    // one SubInfo per id
    idTopics.forEach(
        (id, topics) -> {
          List<String> args = new ArrayList<>();
          WebSocketMessageCallback<?> cb = null;

          for (String topic : topics) {
            int idx = topic.indexOf(':');
            if (idx == -1) { // no arg => treat as "all"
              continue;
            }
            String arg = topic.substring(idx + 1);
            if (!"all".equals(arg)) {
              args.add(arg);
            }
            // pick callback from any topic (they’re identical for this id)
            if (cb == null) {
              Callback holder = topicCbs.get(topic);
              if (holder != null) cb = holder.callback;
            }
          }

          list.add(new SubInfo(topicPrefix, args, cb));
        });
    return list;
  }

  @Value
  private static class Callback {
    WebSocketMessageCallback<?> callback;
    String id;
    String topic;
  }
}
